scarb/src/core/manifest/toml_manifest.rs

```diff
@@ -665,13 +665,31 @@ impl TomlManifest {
             }
         } else if auto_detect {
             // Auto-detect test target.
+            let external_contracts = targets
+                .iter()
+                .filter(|target| target.kind == TargetKind::STARKNET_CONTRACT)
+                .filter_map(|target| target.params.get("build-external-contracts"))
+                .filter_map(|value| value.as_array())
+                .flatten()
+                .filter_map(|value| value.as_str().map(ToString::to_string))
+                .sorted()
+                .dedup()
+                .collect_vec();
             let source_path = self.lib.as_ref().and_then(|l| l.source_path.clone());
             let target_name: SmolStr = format!("{package_name}_unittest").into();
             let target_config = TomlTarget::<TomlExternalTargetParams> {
                 name: Some(target_name),
                 source_path,
-                params: TestTargetProps::default().try_into()?,
+                params: TestTargetProps::default()
+                    .with_build_external_contracts(external_contracts.clone())
+                    .try_into()?,
             };
+            let external_contracts = external_contracts
+                .into_iter()
+                .chain(vec![format!("{package_name}::*")])
```

Comment:
The previous target is a `unit` test one, so the tested package is the main component of the compilation unit. This means, that it will be searched for contracts as a main package (i.e. it's in `test_crate_ids`). 

For the `integration` ones (below),  the tested package is treated as any other dependency, so it's not in `test_crate_ids` and thus would not be searched for contracts, unless we explicitly ask for this. 

We also can assume, that `build-external-contracts` does not include this package already, as it's defined in `starknet-contract` target of said package, thus is `external` to the package in question.

Reply:
Thanks! As a nitpick this code can be placed further down for better readability, for example on line 702.